File: src/application/builders/sequencing_builder.py
from __future__ import annotations

from domain.models import (
    Analysis,
    SamplePreparation,
    Sequencing,
    SequencingData,
    SequencingEntry,
)
from domain.utils import as_list, has_any_keys, resolve_source_id


class SequencingBuilder:
    def build_sequencing_data(
        self, predictive_number: str, payload: dict
    ) -> SequencingData:
        entries_payload = as_list(payload.get("sequencing_entries"))
        if not entries_payload:
            entries_payload = [payload]
        entries: SequencingData = []
        for entry_payload in entries_payload:
            if not isinstance(entry_payload, dict):
                continue
            source_id = resolve_source_id(entry_payload, predictive_number)
            entries.append(
                SequencingEntry(
                    predictive_number=predictive_number,
                    source_id=source_id,
                    fixed_block_identifier=self._build_fixed_block_identifier(
                        entry_payload.get("fixed_block", entry_payload)
                    ),
                    sample_preparation=self._build_sample_preparation(
                        entry_payload.get("sample_preparation", entry_payload)
                    ),
                )
            )
        return entries

    def _build_fixed_block_identifier(self, payload: object) -> str | None:
        if not isinstance(payload, dict) or not payload:
            return None
        return payload.get("block_identifier")

    def _build_sample_preparation(self, payload: object) -> SamplePreparation | None:
        if not isinstance(payload, dict) or not payload:
            return None
        if not has_any_keys(
            payload,
            [
                "sampleprep_identifier",
                "belongs_to_material",
                "input_amount",
                "library_preparation_kit",
                "target_enrichment_kit",
                "sequencing_run",
                "sequencing",
            ],
        ):
            return None
        sequencing_run = None
        direct_run = payload.get("sequencing_run", payload.get("sequencing"))
        if isinstance(direct_run, dict):
            sequencing_run = self._build_sequencing_run(direct_run)
        return SamplePreparation(
            sampleprep_identifier=payload.get("sampleprep_identifier"),
            belongs_to_material=payload.get("belongs_to_material"),
            input_amount=payload.get("input_amount"),
            library_preparation_kit=payload.get("library_preparation_kit"),
            pcr_free=payload.get("pcr_free"),
            target_enrichment_kit=payload.get("target_enrichment_kit"),
            full_sequence_genes=payload.get("full_sequence_genes"),
            partial_sequence_genes=payload.get("partial_sequence_genes"),
            umis_present=payload.get("umis_present"),
            intended_insert_size=payload.get("intended_insert_size"),
            intended_read_length=payload.get("intended_read_length"),
            sequencing=sequencing_run,
        )

    def _build_sequencing_run(self, payload: object) -> Sequencing | None:
        if not isinstance(payload, dict) or not payload:
            return None
        if not has_any_keys(
            payload,
            [
                "sequencing_identifier",
                "belongs_to_sample_preparation",
                "sequencing_date",
                "sequencing_platform",
                "sequencing_instrument_model",
                "sequencing_method",
                "analysis",
            ],
        ):
            return None
        analysis = [
            self._build_analysis(item)
            for item in as_list(payload.get("analysis"))
            if isinstance(item, dict)
        ]
        return Sequencing(
            sequencing_identifier=payload.get("sequencing_identifier"),
            belongs_to_sample_preparation=payload.get("belongs_to_sample_preparation"),
            sequencing_date=payload.get("sequencing_date"),
            sequencing_platform=payload.get("sequencing_platform"),
            sequencing_instrument_model=payload.get("sequencing_instrument_model"),
            sequencing_method=payload.get("sequencing_method"),
            median_read_depth=payload.get("median_read_depth"),
            observed_read_length=payload.get("observed_read_length"),
            observed_insert_size=payload.get("observed_insert_size"),
            percentage_q30=payload.get("percentage_q30"),
            percentage_tr20=payload.get("percentage_tr20"),
            other_quality_metrics=payload.get("other_quality_metrics"),
            analysis=analysis[0] if analysis else None,
        )

    def _build_analysis(self, payload: dict) -> Analysis:
        return Analysis(
            analysis_identifier=payload.get("analysis_identifier"),
            belongs_to_sequencing=payload.get("belongs_to_sequencing"),
            physical_data_location=payload.get("physical_data_location"),
            abstract_data_location=payload.get("abstract_data_location"),
            data_formats_stored=payload.get("data_formats_stored"),
            algorithms_used=payload.get("algorithms_used"),
            reference_genome_used=payload.get("reference_genome_used"),
            bioinformatic_protocol_used=payload.get("bioinformatic_protocol_used"),
            bioinformatic_protocol_deviation=payload.get(
                "bioinformatic_protocol_deviation"
            ),
            reason_for_bioinformatic_protocol_deviation=payload.get(
                "reason_for_bioinformatic_protocol_deviation"
            ),
            wgs_guideline_followed=payload.get("wgs_guideline_followed"),
        )

File: src/Uploader/Uploader.Application/Builders/SequencingBuilder.cs
using System.Text.Json.Nodes;
using Uploader.Application.Json;
using Uploader.Domain;

namespace Uploader.Application.Builders;

/// <summary>Builds the sequencing pipeline (entries -> sample prep -> sequencing -> analysis).</summary>
public sealed class SequencingBuilder
{
    public IReadOnlyList<SequencingEntry> BuildSequencingData(string predictiveNumber, JsonObject payload)
    {
        var entriesPayload = RawJson.AsList(payload["sequencing_entries"]);
        if (entriesPayload.Count == 0)
        {
            entriesPayload = [payload];
        }

        var entries = new List<SequencingEntry>();
        foreach (var node in entriesPayload)
        {
            if (node is not JsonObject entry)
            {
                continue;
            }

            entries.Add(new SequencingEntry
            {
                PredictiveNumber = predictiveNumber,
                SourceId = RawJson.ResolveSourceId(entry, predictiveNumber),
                FixedBlockIdentifier = BuildFixedBlockIdentifier(GetOrSelf(entry, "fixed_block")),
                SamplePreparation = BuildSamplePreparation(GetOrSelf(entry, "sample_preparation")),
            });
        }

        return entries;
    }

    private static JsonObject? GetOrSelf(JsonObject entry, string key) =>
        (entry.ContainsKey(key) ? entry[key] : entry) as JsonObject;

    private static string? BuildFixedBlockIdentifier(JsonObject? payload) =>
        payload is null || payload.Count == 0 ? null : payload.GetString("block_identifier");

    private static SamplePreparation? BuildSamplePreparation(JsonObject? payload)
    {
        if (payload is null || payload.Count == 0)
        {
            return null;
        }

        if (!RawJson.HasAnyKeys(
            payload,
            "sampleprep_identifier",
            "belongs_to_material",
            "input_amount",
            "library_preparation_kit",
            "target_enrichment_kit",
            "sequencing_run",
            "sequencing"))
        {
            return null;
        }

        var directRun = RawJson.FirstDict(payload["sequencing_run"] ?? payload["sequencing"]);

        return new SamplePreparation
        {
            SampleprepIdentifier = payload.GetString("sampleprep_identifier"),
            BelongsToMaterial = payload.GetString("belongs_to_material"),
            InputAmount = payload.GetString("input_amount"),
            LibraryPreparationKit = payload.GetString("library_preparation_kit"),
            PcrFree = payload.GetBool("pcr_free"),
            TargetEnrichmentKit = payload.GetString("target_enrichment_kit"),
            FullSequenceGenes = payload.GetStringList("full_sequence_genes"),
            PartialSequenceGenes = payload.GetStringList("partial_sequence_genes"),
            UmisPresent = payload.GetBool("umis_present"),
            IntendedInsertSize = payload.GetInt("intended_insert_size"),
            IntendedReadLength = payload.GetInt("intended_read_length"),
            Sequencing = directRun is null ? null : BuildSequencingRun(directRun),
        };
    }

    private static Sequencing? BuildSequencingRun(JsonObject? payload)
    {
        if (payload is null || payload.Count == 0)
        {
            return null;
        }

        if (!RawJson.HasAnyKeys(
            payload,
            "sequencing_identifier",
            "belongs_to_sample_preparation",
            "sequencing_date",
            "sequencing_platform",
            "sequencing_instrument_model",
            "sequencing_method",
            "analysis"))
        {
            return null;
        }

        var analysis = RawJson.AsList(payload["analysis"])
            .OfType<JsonObject>()
            .Select(BuildAnalysis)
            .ToList();

        return new Sequencing
        {
            SequencingIdentifier = payload.GetString("sequencing_identifier"),
            BelongsToSamplePreparation = payload.GetString("belongs_to_sample_preparation"),
            SequencingDate = payload.GetString("sequencing_date"),
            SequencingPlatform = payload.GetString("sequencing_platform"),
            SequencingInstrumentModel = payload.GetString("sequencing_instrument_model"),
            SequencingMethod = payload.GetString("sequencing_method"),
            MedianReadDepth = payload.GetInt("median_read_depth"),
            ObservedReadLength = payload.GetInt("observed_read_length"),
            ObservedInsertSize = payload.GetInt("observed_insert_size"),
            PercentageQ30 = payload.GetDouble("percentage_q30"),
            PercentageTr20 = payload.GetDouble("percentage_tr20"),
            OtherQualityMetrics = payload.GetString("other_quality_metrics"),
            Analysis = analysis.Count > 0 ? analysis[0] : null,
        };
    }

    private static Analysis BuildAnalysis(JsonObject payload) => new()
    {
        AnalysisIdentifier = payload.GetString("analysis_identifier"),
        BelongsToSequencing = payload.GetString("belongs_to_sequencing"),
        PhysicalDataLocation = payload.GetString("physical_data_location"),
        AbstractDataLocation = payload.GetString("abstract_data_location"),
        DataFormatsStored = payload.GetStringList("data_formats_stored"),
        AlgorithmsUsed = payload.GetStringList("algorithms_used"),
        ReferenceGenomeUsed = payload.GetString("reference_genome_used"),
        BioinformaticProtocolUsed = payload.GetString("bioinformatic_protocol_used"),
        BioinformaticProtocolDeviation = payload.GetString("bioinformatic_protocol_deviation"),
        ReasonForBioinformaticProtocolDeviation = payload.GetString("reason_for_bioinformatic_protocol_deviation"),
        WgsGuidelineFollowed = payload.GetString("wgs_guideline_followed"),
    };
}
